Passive scalar support

gloopydoop opened the pull request:
This PR contains:

-  Adjoint passive scalar support
- A scalar mixing objective function
- An example of a mixer

gloopydoop committed:
little adjustment

gloopydoop committed:
Update sources/CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Declare in same order as usage in the files

timfelle committed:
Move MMA settings into the solver section

timfelle committed:
Small clean up of optimizer

timfelle committed:
Check against the result folder instead of a log

timfelle committed:
Add a factory for generic optimizer

timfelle committed:
Fix uninitialized variables

timfelle committed:
Update driver with the parameter read

timfelle committed:
Simplify optimizer (Invalidated previous results)

timfelle committed:
Updates are functional

timfelle committed:
Slight cleanup

timfelle committed:
Add base initializer

timfelle committed:
Updates to our interfaces

timfelle committed:
Align the interfaces with style from neko

timfelle committed:
Improve the driver structure.

timfelle committed:
Use the simulation reset function

timfelle committed:
Update parameter order

timfelle committed:
Add a file reader function for json

gloopydoop committed:
Update sources/neko_ext/math/bcknd/device/cuda/math_ext_kernel.h

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Update sources/problems/objectives/lube_term_objective.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
PR suggestions

gloopydoop committed:
Merge branch 'GPU_support/objective_function' of github.com:ExtremeFLOW/neko-top into GPU_support/objective_function

AbbasMoosaV committed:
Restructure mma_t removing the mma_factory and utilizing submodules for the backend. Cuda and CPU backends checked for example/mma. Further cleanups needed.

gloopydoop committed:
PR comments

gloopydoop committed:
Update sources/problems/objectives/lube_term_objective.f90

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Apply suggestions from code review

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
import c_ptr, c_int

gloopydoop committed:
Merge branch 'GPU_support/objective_function' of github.com:ExtremeFLOW/neko-top into GPU_support/objective_function

gloopydoop committed:
fromatting

gloopydoop committed:
slight updates

gloopydoop committed:
minor changes

gloopydoop committed:
csv_file_t

gloopydoop committed:
PR comments

gloopydoop committed:
lint

gloopydoop committed:
PR comments

gloopydoop committed:
.txt -> .csv

gloopydoop committed:
formatting

gloopydoop committed:
back to the zeroth iteration

gloopydoop committed:
Update sources/neko_ext/math/bcknd/device/cuda/math_ext_kernel.h

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
initial condition

timfelle committed:
Improved logging (#91)

I guess we discussed this a bit this morning, but this is what I had in
mind.
I'm putting this in draft mode because I want comments before I move any
further with this:

- Do we think we'll want this in other optimization? ie, would a
`logger_t` be useful?
- Are we all happy with enforcing a log name to all
constraints/objectives so we can write them out?

But perhaps more importantly, what I've written currently is pretty
susceptible to truncation of that long string (buffer? is this the
correct word? that was the intention of the variable name `buff`)

### Example output

```
n =     86400, m =         1, tolerance =  1.00000000000000002E-03
iter, Total objective function, Dissipation, out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
  0,     5.3696209634E+01,     4.0441625841E+01,     1.3254583793E+01,    -3.4113716238E-01,     1.7976931349+308,     1.7976931349+308,     1.0000000000E+00
  1,     5.3631210218E+01,     4.0441625841E+01,     1.3189584378E+01,    -3.4067827356E-01,     3.4063388244E-01,     1.8251180684E+00,     1.0000000000E+00
  2,     5.3266158113E+01,     4.0127814834E+01,     1.3138343279E+01,    -3.4021952588E-01,     3.4017508569E-01,     5.0791113981E+00,     1.0000000000E+00
  3,     5.2946050791E+01,     3.9894493099E+01,     1.3051557691E+01,    -3.3973026153E-01,     3.3968611728E-01,     4.1537413064E+00,     1.0000000000E+00
  4,     5.2488335978E+01,     3.9561169448E+01,     1.2927166530E+01,    -3.3922171690E-01,     3.3917795103E-01,     5.6123647507E+00,     1.0000000000E+00
  5,     5.1825567047E+01,     3.9082081599E+01,     1.2743485448E+01,    -3.3889046816E-01,     3.3884813432E-01,     7.9908112190E+00,     1.0000000000E+00
  6,     5.1000305086E+01,     3.8491289553E+01,     1.2509015532E+01,    -3.3852679847E-01,     3.3848677945E-01,     9.8497959414E+00,     1.0000000000E+00
  7,     5.0354702281E+01,     3.8025030555E+01,     1.2329671726E+01,    -3.3866646158E-01,     3.3862891930E-01,     7.8845237703E+00,     1.0000000000E+00
  8,     4.9785292383E+01,     3.7615237248E+01,     1.2170055135E+01,    -3.3902719413E-01,     3.3899308096E-01,     7.5623243319E+00,     1.0000000000E+00

```

### Example output using `csv_file_type`
```
iter, Total objective function, Dissipation, Out of plane stresses, Volume constraint, KKTmax, KKTnorm2, scalaing factor
0.0000000000000000,53.696209634065141,40.441625840817053,13.254583793248088,-0.34113716238378983,0.0000000000000000,0.0000000000000000,1.0000000000000000
1.0000000000000000,53.631210218451116,40.441625840817053,13.189584377634059,-0.34067827356064134,0.34063388243730330,1.8251180684119177,1.0000000000000000
2.0000000000000000,53.266158113095081,40.127814833617478,13.138343279477603,-0.34021952587850796,0.34017508568651300,5.0791113980976883,1.0000000000000000
3.0000000000000000,52.946050790610592,39.894493099386999,13.051557691223593,-0.33973026152689706,0.33968611727815884,4.1537413063900503,1.0000000000000000
4.0000000000000000,52.488335977948438,39.561169447724346,12.927166530224092,-0.33922171690237829,0.33917795102621462,5.6123647507375267,1.0000000000000000
5.0000000000000000,51.825567047220396,39.082081599088639,12.743485448131759,-0.33889046815908813,0.33884813432446081,7.9908112190203324,1.0000000000000000
6.0000000000000000,51.000305085635944,38.491289553207920,12.509015532428020,-0.33852679846645700,0.33848677945129146,9.8497959414055956,1.0000000000000000
7.0000000000000000,50.354702281281916,38.025030555367294,12.329671725914622,-0.33866646158438390,0.33862891930457578,7.8845237703490136,1.0000000000000000
8.0000000000000000,49.785292383215989,37.615237248149093,12.170055135066898,-0.33902719412641674,0.33899308096021136,7.5623243318792319,1.0000000000000000
```

gloopydoop committed:
merge develop

timfelle committed:
Merge commit '3f11b506ff4b802167a10aafb58841c947ea526f' into driver/json

timfelle committed:
Merge commit '7a0b303bcd03c05e12edd54a976826a89824f2b2' into driver/json

AbbasMoosaV committed:
Cleanups and updates based on PR comments.

gloopydoop committed:
Apply suggestions from code review

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
PR comments

timfelle committed:
Apply suggestions from code review

AbbasMoosaV committed:
Update the structure of mma.cu.

timfelle committed:
Fix the possibility to specify the number of processors to run with

timfelle committed:
Allow specification of processor count for execution (#92)

Introduce the ability to specify the number of processors to use via the
`NPROCS` environment variable, enhancing flexibility in resource
allocation during script execution.

AbbasMoosaV committed:
Update the structure of mma_kernel.h

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into GPU_support/objective_function

gloopydoop committed:
Merge branch 'GPU_support/objective_function' of github.com:ExtremeFLOW/neko-top into GPU_support/objective_function

gloopydoop committed:
mma

timfelle committed:
Enhance JSON handling and driver structure improvements (#72)

Introduce json based initialization for all main types.

Additionally I have split the simulation out of the problem in order to
simplify dependencies.

Depends on #76

timfelle committed:
Merge branch 'develop' into GPU_support/objective_function

timfelle committed:
Fix formatting

timfelle committed:
Up the linting score a bit in general

timfelle committed:
Gpu support (#79)

Originally this was 5 branches, and I wanted to submit 5 individual PRs
so this would be a bit cleaner.
But essentially this is bringing the Rugby ball case up to date with GPU
support.
The three branches were:
`GPU_support/masks`
`GPU_support/volume_constraint`
`GPU_support/objective_function`
`GPU_support/design` (which also contained the RAMP mappings etc)
`GPU_support/MMA_on_CPU` This one won't be needed, but it essentially
mem_copied everything from device to host, did MMA on the CPU, then
copied back.

But I was able to run the rugby ball on GPU so I was quite happy!
However, the optimization history wasn't EXACTLY the same as CPUs, so
I'm keeping this on draft until I can test more and resolve this.

**EDIT**

Ok I'm more and more confident the problem is with MMA on CPU, so I
think we can take this off draft mode. Also it looks like PR#85 needs a
lot of revisions and I want to get this PR through so we can all start
testing things on GPU. I apologies if there is bug is in here unrelated
to the hacky memcopies I performed.

 **note:**
I made a rather fundamental change to PDE filter, ie, solving for
$$\Delta \tilde{\rho}$$ not $$\tilde{\rho}$$, so already we won't match
with develop exactly, due to the residual prescribed by the PDE filter.

### Convergence history on develop
```
  0,  5.36962096340651414E+01, -3.41137162383789827E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962096340651414E+01, -3.40678273560641343E-01,  3.40633882437303304E-01,  1.82511436798967752E+00,  1.00000000000000000E+00
  2,  5.34909202977577962E+01, -3.40219526392907701E-01,  3.40175086197066656E-01,  2.93977673685759111E+00,  1.00000000000000000E+00
  3,  5.32662799775875442E+01, -3.39695799113361951E-01,  3.39651438883154300E-01,  3.19543034281953853E+00,  1.00000000000000000E+00
  4,  5.29593114254205517E+01, -3.39109389875194511E-01,  3.39065248562431587E-01,  3.98656352093350419E+00,  1.00000000000000000E+00
  5,  5.25307111679251335E+01, -3.38585236396372891E-01,  3.38541901793160127E-01,  5.22808835345366951E+00,  1.00000000000000000E+00
  6,  5.19234539710996330E+01, -3.38143391208646948E-01,  3.38102175627622070E-01,  7.03658279719623980E+00,  1.00000000000000000E+00
  7,  5.11597748945414921E+01, -3.37928412343896978E-01,  3.37888261694012759E-01,  8.82061398036337785E+00,  1.00000000000000000E+00
  8,  5.05602700893721249E+01, -3.38296226436742675E-01,  3.38259998372253234E-01,  6.98241437563093026E+00,  1.00000000000000000E+00
  9,  5.00465163087089451E+01, -3.38384508012592067E-01,  3.38349557059135442E-01,  6.54315441819684906E+00,  1.00000000000000000E+00
 10,  4.95253467948585779E+01, -3.38395758796641910E-01,  3.38362432950295389E-01,  6.75821664949387291E+00,  1.00000000000000000E+00
```

### Convergence history on CPU (with the new changes)
```
  0,  5.36962098917103532E+01, -3.41137162383789827E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962098917103532E+01, -3.40678272361158607E-01,  3.40633881225604174E-01,  1.82511372436923880E+00,  1.00000000000000000E+00
  2,  5.34909201511379564E+01, -3.40219524004039708E-01,  3.40175083796164401E-01,  2.93978525062977747E+00,  1.00000000000000000E+00
  3,  5.32662797551102756E+01, -3.39695795330679007E-01,  3.39651435093588583E-01,  3.19540660892341366E+00,  1.00000000000000000E+00
  4,  5.29593111639978673E+01, -3.39109383832420674E-01,  3.39065242501398190E-01,  3.98658874364438764E+00,  1.00000000000000000E+00
  5,  5.25307107108303555E+01, -3.38585227652321519E-01,  3.38541893041661768E-01,  5.22806768673648037E+00,  1.00000000000000000E+00
  6,  5.19234524546626943E+01, -3.38143379103141128E-01,  3.38102163503164688E-01,  7.03658086336280775E+00,  1.00000000000000000E+00
  7,  5.11597740019539202E+01, -3.37928397133592295E-01,  3.37888246463098230E-01,  8.82063926519529318E+00,  1.00000000000000000E+00
  8,  5.05602689953278457E+01, -3.38296209449853336E-01,  3.38259981376574370E-01,  6.98237874556235294E+00,  1.00000000000000000E+00
  9,  5.00465151924080729E+01, -3.38384488855007104E-01,  3.38349537884281071E-01,  6.54318127887768597E+00,  1.00000000000000000E+00
 10,  4.95253453940975277E+01, -3.38395744127610365E-01,  3.38362418271727805E-01,  6.75818420810124021E+00,  1.00000000000000000E+00
```

### Convergence history on GPU (with the new changes)
```
  0,  5.36962098917055499E+01, -3.41137162383794046E-01,  1.79769313486231571+308,  1.79769313486231571+308,  1.00000000000000000E+00
  1,  5.36962098917055499E+01, -3.40678272944622995E-01,  6.26706799280533833E+00,  6.43310832024656179E+02,  1.00000000000000000E+00
  2,  5.34910118155738559E+01, -3.40219525367835784E-01,  6.26705020196007023E+00,  6.43309822428294638E+02,  1.00000000000000000E+00
  3,  5.32664717580504288E+01, -3.39691202134478287E-01,  6.23507603889123274E+00,  6.41924745328890140E+02,  1.00000000000000000E+00
  4,  5.30975501242981096E+01, -3.39151984540635609E-01,  6.19951479473746137E+00,  6.40112443825842547E+02,  1.00000000000000000E+00
  5,  5.25350413808810828E+01, -3.38644965953474764E-01,  6.16274815428317435E+00,  6.38611032141696114E+02,  1.00000000000000000E+00
  6,  5.19248549714847698E+01, -3.38207041002288433E-01,  6.08546842737498661E+00,  6.34741181740812067E+02,  1.00000000000000000E+00
  7,  5.11622343791975069E+01, -3.37997668480812863E-01,  5.98907463234943904E+00,  6.30183729381854391E+02,  1.00000000000000000E+00
  8,  5.05718316461075759E+01, -3.38299594367291456E-01,  5.86614026930082400E+00,  6.24306904330888301E+02,  1.00000000000000000E+00
  9,  5.00919821259184630E+01, -3.38447606977853088E-01,  5.77873574707956283E+00,  6.19710815500791000E+02,  1.00000000000000000E+00
 10,  4.96736731441919943E+01, -3.38192325940090688E-01,  5.69855410966845621E+00,  6.15708752891213408E+02,  1.00000000000000000E+00
```

timfelle committed:
Add missing flag to cmake.

Lost when Testing config was added

timfelle committed:
Build the examples from the runner script

timfelle committed:
Fix a few things in the verification run script

timfelle committed:
Move the source term to design.

timfelle committed:
Print MPI safe

timfelle committed:
Move mapping to parent design

timfelle committed:
Move computations to problem

timfelle committed:
Move design output to design.

timfelle committed:
Move the last pieces out of the steady state problem, hence it is removed.

timfelle committed:
revert change

timfelle committed:
Update sources/optimizer/mma_optimizer.f90

gloopydoop committed:
missing import

gloopydoop committed:
half way there

gloopydoop committed:
Merge branch 'bugfix/memcopy' into fix/rugby_ball_case

timfelle committed:
missing import in `mma_optimizer` (#96)

missing import

gloopydoop committed:
and the driver too

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into fix/rugby_ball_case

AbbasMoosaV committed:
Merge develop into feature/gpu_support_MMA

AbbasMoosaV committed:
Add error for unknown backend in mma_init_attribute based on PR comments.

timfelle committed:
Fix verification run script and build examples from runner script (#94)

Streamline the execution process by building example executables from
the runner script.

Update the verification example. New filenames were not used properly.
Also the example will now check against the result folder, rather than a
manually added reference.

Also added an MPI guard on a few print statements

timfelle committed:
Bring rugby ball up to date (#97)

Here I've just updated the case file and driver of `rugby_ball` to
reflect the to updates in `rugby_verification`

**NOTE**

This will fail on device at the 2nd iteration due to the KKT check at
line 413 of `mma.f90`. Substituting
```fortran
    if (NEKO_BCKND_DEVICE .eq. 0) then
       call mma_KKT_cpu(this, x, df0dx, fval, dfdx)
    else
       write(stderr, *) "Device not supported for MMA."
       error stop
    end if
```

with
``` fortran
    if (this%backend == 'cpu') then
       call mma_KKT_cpu(this, x, df0dx, fval, dfdx)
    else
       write(stderr, *) "Device not supported for MMA."
       error stop
    end if
```

Is how PR#79 was run

AbbasMoosaV committed:
Updates based on suggested improvments (Linting).

AbbasMoosaV committed:
Improving the format.

AbbasMoosaV committed:
Resolve compiler warnings.

AbbasMoosaV committed:
Merge develop and resolve conflicts.

AbbasMoosaV committed:
Correct the (only :) typo in all files in the source folder.

AbbasMoosaV committed:
Linting updates.

AbbasMoosaV committed:
More linting updates.

AbbasMoosaV committed:
Linting!

timfelle committed:
Formatting

AbbasMoosaV committed:
Merge mma%free%device and mma%free%cpu to mma%free.

gloopydoop committed:
time to test the linting

gloopydoop committed:
linting

gloopydoop committed:
linting

gloopydoop committed:
forgot the cmake

AbbasMoosaV committed:
Updates based on PR comments 1.

timfelle committed:
Linter updates

timfelle committed:
fix bug

gloopydoop committed:
up the linting

gloopydoop committed:
maybe the docs?

timfelle committed:
Up the score

timfelle committed:
upgrade flinter for the CI

timfelle committed:
fix requirements

timfelle committed:
revert changes to improve compatibility with other PR's. Simplifying the merge queue

gloopydoop committed:
suggestion for field writing

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into refactor/design-problem

gloopydoop committed:
merge develop and name change

timfelle committed:
Refactor/design-problem (#93)

This PR build on the structure.

The design, problem and simulation is untangled. The steady state
problem are no longer needed.

gloopydoop committed:
u_adj

gloopydoop committed:
unused import

gloopydoop committed:
Merge branch 'docs/adjoint' of github.com:ExtremeFLOW/neko-top into docs/adjoint

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into docs/adjoint

AbbasMoosaV committed:
Updates based on PR comments 2.

gloopydoop committed:
links

gloopydoop committed:
Update sources/CMakeLists.txt

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Updates based on PR comments 3.

AbbasMoosaV committed:
Updates based on the PR comments 4.

AbbasMoosaV committed:
Updates based on PR comments 5.

gloopydoop committed:
forgot to add adjoint.md

gloopydoop committed:
Merge branch 'docs/adjoint' of github.com:ExtremeFLOW/neko-top into docs/adjoint

gloopydoop committed:
big equations good, inline equations bad

gloopydoop committed:
few!

timfelle committed:
Documentation for adjoint fluid + name changes (#99)

The PR is moving in the direction of getting the adjoint passive scalar
off the ground. It consists of:
1) A bit of documentation for the adjoint velocity equations (I'll fill
in the adjoint passive scalar in a future PR). It isn't linked to the
rest of the docs, I'll wait for PR#95.
2) A lot of name changes such that `adjoint` becomes `adjoint_fluid`
where applicable, opening the way for us to have
- `fluid` and `adjoint_fluid`
- `scalar` and `adjoint_scalar`

gloopydoop committed:
starting

timfelle committed:
Revert "Documentation for adjoint fluid + name changes (#99)"

This reverts commit e86bc8fcf9e83556e64f4769573ae9840bb8cadf, reversing
changes made to 4a78019056cd8d11e5e3a9b0030b3ee287e08f07.

gloopydoop committed:
halfway

timfelle committed:
formatting

timfelle committed:
Revert documentation changes and update adjoint references (#100)

Revert previous documentation changes related to adjoint fluid and
update references to use a unified adjoint scheme. Adjustments include
modifications to various modules and removal of outdated documentation.

gloopydoop committed:
wooooowwwwww that was close

gloopydoop committed:
these guys were in comments, but still to be safe

AbbasMoosaV committed:
Merge in develop and resolve conflicts.

AbbasMoosaV committed:
Updates based on PR comments 6.

AbbasMoosaV committed:
Little formatting.

AbbasMoosaV committed:
Format!

AbbasMoosaV committed:
Formatting!

AbbasMoosaV committed:
Formatting!

gloopydoop committed:
Revert "Revert documentation changes and update adjoint references"

gloopydoop committed:
Merge commit 'cb271fb40e194239329387628bee1a79d98daed9' into revert-100-revert/adjoint_update

gloopydoop committed:
Merge commit '875569046c8b2edf384e2660dd547adcb06a2b2c' into revert-100-revert/adjoint_update

gloopydoop committed:
linting

gloopydoop committed:
linting

timfelle committed:
update cleanup function

timfelle committed:
Apply suggestion

timfelle committed:
this is what chat-GPT says to do :sweat_smile:  (#102)

what a kerfuffle...

according to chatGPT:

If you reverted the original PR using another PR, GitHub provides a
"Revert" button:

Go to the PR that reverted your original changes.
Click on the "Revert" button (if available). This will create a new PR
that restores the original changes.
Then, apply your fix on top of that before merging.

I hope that's what has been done in this PR, but please double check

timfelle committed:
Merge remote-tracking branch 'origin/develop' into feature/gpu_support_MMA

timfelle committed:
Lint

timfelle committed:
lint

timfelle committed:
Feature/gpu support mma (#85)

This updates the mma_optimizer structure, introduces mma_factory for CPU
and device backend types and generally, adds CUDA support for mma
solver.

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into feature/clean_passive_scalar

gloopydoop committed:
adjoint scalar scheme and pnpn

gloopydoop committed:
backup

timfelle committed:
Make sure to not use built-in names for variables

timfelle committed:
Exit with an error if examples cannot be compiled

gloopydoop committed:
WIP

timfelle committed:
Avoid built-in names and handle compilation errors (#104)

Ensure variable names do not conflict with built-in names and exit with
an error message if examples fail to compile.

gloopydoop committed:
compiling, still need to adjust examples

gloopydoop committed:
compiling, no case file yet

gloopydoop committed:
example case

gloopydoop committed:
Merge remote-tracking branch 'origin/develop' into feature/clean_passive_scalar

gloopydoop committed:
WIP

gloopydoop committed:
small change

gloopydoop committed:
strange line

gloopydoop committed:
still that strange insertion

gloopydoop committed:
there's something wrong the JSON, trying a rebuild

timfelle committed:
really small change (#108)

I think the changes are obvious.
But I'd never seen that `"output_control": "never"` before. That's
really nice for us :+1:

gloopydoop committed:
WIP

gloopydoop committed:
finally started working! Time for ALOT of cleaning

gloopydoop committed:
let the cleaning begin

gloopydoop committed:
findent adjoint

gloopydoop committed:
findent source_terms

gloopydoop committed:
WIP

gloopydoop committed:
this should be broken into two

gloopydoop committed:
should be adifferent PR

gloopydoop committed:
removed

gloopydoop committed:
Merge branch 'fix/adjoint_premultiply' into feature/clean_passive_scalar

gloopydoop committed:
Merge branch 'fix/adapt_to_neko' into feature/clean_passive_scalar

gloopydoop committed:
Update for `neko` event (#111)

Update based on `neko` PR #1736

gloopydoop committed:
remove premultiply adjoint forcing with mass matrix (#112)

The pre-multiplication is handled in the source term handler, we
shouldn't have it here.

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

AbbasMoosaV committed:
Solve the mpi reduce issue in subsolve in mma_device. (#113)

Small fix in  mma_device.

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
WIP

timfelle committed:
Refactor the design_t (#103)

Major update to the design type.
This PR will improve the reuseability of the design type by hiding
implementation details behind a generic API.

gloopydoop committed:
WIP

timfelle committed:
Sync formatting to neko standards (#125)

Update formatting commands to align with neko's specifications.

Continuation lines starting with a `&` is not indented to 0 compared to
their current scope.

This is commonly used to break long strings.

gloopydoop committed:
WIP

gloopydoop committed:
Fix fluid ICs (#126)

This is in response to my `neko` PR passing the subdict to ICs.

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Align with new KSP in neko (#129)

Update the usage of the Krylov solver to align with the new KSP
interface, replacing calls to `krylov_solver_destroy` with the
appropriate `free` method for memory management.

timfelle committed:
Cleanup and fix examples in the repository (#128)

Remove outdated and broken example files while fixing existing examples
to ensure proper functionality. This cleanup enhances the overall
quality and maintainability of the codebase.

timfelle committed:
Documentation improvements and example updates (#95)

Enhance documentation clarity and structure by updating references,
fixing formatting issues, and removing redundant content. Add missing
examples and address warnings in the documentation.

gloopydoop committed:
merged develop

gloopydoop committed:
time to lint

gloopydoop commented:
This PR is getting too big and really should have been many little PRs, so I'm getting it ready for review before I continue in a separate PR.
Essentially, now, the passive_scalar examples "works" in the sense that it runs, but it does "work" in the sense that we get a nice results. The intention will be to merge this PR with #124 and #127 such that everything is in the case file, then tinker with the parameters a little to get the example nice. This will be done in a separate PR.

gloopydoop committed:
cleaning

gloopydoop committed:
cleaning

gloopydoop committed:
cleaning

gloopydoop committed:
cleaning

gloopydoop committed:
cleaning

gloopydoop committed:
little one

gloopydoop committed:
removing steady_simcomp changes for a different PR

gloopydoop committed:
revert steady simcomp to develop

gloopydoop committed:
cleaning

gloopydoop committed:
formatting

gloopydoop committed:
unused variables

gloopydoop committed:
control_flow deleted

gloopydoop committed:
unused variables

gloopydoop committed:
unused variables

timfelle committed:
Align with new checkpoint structure in Neko (#133)

Update the code to align with the new checkpoint structure, ensuring
that time-related parameters are accessed correctly throughout the
adjoint and reset subroutines.

gloopydoop committed:
adapt to time controller

timfelle committed:
Ensure cross compilation support and clean up build flags (#122)

Introduce a new file for compiler flags to streamline the build process
and ensure compatibility with cross compilation. This change simplifies
the management of compiler options for different build types.

timfelle committed:
Fix aliasing module references (#123)

Update module references from `simulation` to `simulation_m` to resolve
aliasing issues. Adjust initialization and cleanup processes
accordingly.

Fixes: #116

gloopydoop committed:
Definition of steady residual + logging (#131)

I'm trying to dissect some of the changes that were originally in #110 .

Here I've changed the definition of the steady residual and added a
little csv writer so it can be monitored

timfelle commented on examples/passive_scalar/user.f90:
```diff
@@ -0,0 +1,107 @@
+       s%x(i,1,1,1) = L / (1.0_rp + exp(-k*(s%dof%z(i,1,1,1) - z_0)))
+    end do
+
+    !@ todo
+    ! I haven't checked if I need a memcopy here, but the cyl_boundary_layer
+    ! does do one, so I'm assuming it's not needed.
```
If you want this to be supported for GPU you do need it.

gloopydoop replied:
turns out you don't, it's handled in `set_scalar_ic_common`, but I'll get rid of this todo :+1:

timfelle commented on sources/problems/objectives/scalar_mixing_objective_function.f90:
```diff
@@ -0,0 +1,290 @@
+  use utils, only: neko_error
+  use adjoint_mixing_scalar_source_term, only: &
+       adjoint_mixing_scalar_source_term_t
+  ! delete after
```
After what?

gloopydoop replied:
after the simulation takes care of $\mathbf{u} \cdot \mathbf{u}^\dagger$

timfelle commented on sources/problems/objectives/minimum_dissipation_objective.f90:
```diff
@@ -131,7 +130,7 @@ module minimum_dissipation_objective
   !> The common constructor using a JSON object.
   !! @param design the design.
-  !! @param fluid the fluid scheme.
+  !! @param fluid the fluid fluid_adj.
```
did you mean to replace this

timfelle commented on sources/problems/objectives/minimum_dissipation_objective.f90:
```diff
@@ -211,7 +210,7 @@ end subroutine minimum_dissipation_free
   !> Compute the objective function.
   !! @param design the design.
-  !! @param fluid the fluid scheme.
+  !! @param fluid the fluid fluid_adj.
```
same here

timfelle commented on sources/simulation/simulation.f90:
```diff
@@ -58,7 +58,7 @@ module simulation
      !> The fluid
      class(fluid_scheme_incompressible_t), public, pointer :: &
-          fluid_scheme => null()
+          fluid => null()
```
How is this relevant for the adjoint scalar?

gloopydoop replied:
I would probably make the argument that since we have `neko_case` and `adjoint_case` there's probably no need for the simulation to have an additional pointer for a `fluid`, the same could be achieved with some associate statements when initializing the sampler.

But I guess this a prime example of our inconsistency in naming convention :sweat_smile:

but it goes `case%fluid` and `case%scalar` not `case%fluid_scheme`, hence the change.

But a few lines below has dreadful inconsistency :laughing: 

`this%adjoint_case%fluid_adj%p_adj`
should really be
`this%adjoint_case%adjoint_fluid%adjoint_p`

gloopydoop replied:
now we have `fluid` `adjoint_fluid` `scalar` and `adjoint_scalar`

After all this is through I'll do another PR with a lot of `fluid_adj` => `adjoint_fluid` in other sections of the code

timfelle commented on sources/simulation/simulation.f90:
```diff
@@ -94,15 +94,18 @@ subroutine simulation_init(this, parameters)
 
     ! initialize the primal
     call neko_init(this%neko_case)
+
```
Why the extra line break

timfelle commented on sources/simulation/simulation.f90:
```diff
@@ -94,15 +94,18 @@ subroutine simulation_init(this, parameters)
-       this%fluid_scheme => fluid
+       this%fluid => fluid
 
     end select
 
+
+
```
Why introduce two empty lines here?

timfelle commented on sources/adjoint/adjoint_case.f90:
```diff
@@ -76,27 +88,34 @@ subroutine adjoint_init_from_json(this, neko_case)
     type(case_t), target, intent(inout) :: neko_case
     real(kind=rp) :: tol
+    logical :: have_scalar
+
+
 
```
Why 3 empty lines?

timfelle reviewed (changes requested):
Are you certain this PR is ready for review? There are tons of todos, empty lines and comments that seem more like debugging than something that should be part of the code.

gloopydoop committed:
cleaning todo

gloopydoop committed:
todo on delete after

gloopydoop committed:
copy and past

gloopydoop committed:
unused files

timfelle reviewed (changes requested):
When running the example i get the following:
 *** ERROR: Invalid user defined inflow condition ***

gloopydoop committed:
WIP

gloopydoop committed:
Mapping cascade (#124)

This PR has a couple of nice features, but is still a bit of a work in
progress (maybe)

- The main contribution is the notion of a mapping cascade and a lot of
JSON support for the pre-existing mappings. Now we prescribe a list of
mappings which are performed sequentially, and it also takes care of the
chain rule moving backwards.
- There are some cosmetic changes to variable names, particularly
`sens_in` and `sens_out` :sweat_smile:
- Obviously much of the mapping was inspired by the source term handler,
so in principal this could be extended to allow user defined mappings in
some way, but this is to be discussed.

This PR is still in draft mode because I think we should really discuss
who owns the mapping, and how we should handle mapping to multiple
coeffients, and how we should hand constraints defined on intermediate
steps (ie, the potential for the volume constraint to defined on the
filtered design variable)

There are still a few cosmetics changes that need to be done so I'm
leaving this on draft mode, but take a look if you can and tell me what
you think

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

gloopydoop committed:
Optimization domain in JSON (#139)

Wow I can't believe this one was still hardcoded :sweat_smile: 
I've put the "help wanted" flag on because I don't know how this
translates into an optimization problem without a simulation. I guess
the word "domain" for topology optimization means a spatial domain, so
to me I think it should probably be rooted in
`optimization.design.domain`

But it's worth discussing.

Also aiming for many small PRs moving forward :sweat_smile:

---------

Co-authored-by: Tim Felle Olsen <tife@dtu.dk>

timfelle committed:
Improve documentation and fix formatting issues (#138)

Enhance the documentation configuration and deployment, correct markdown
formatting, and complete doxygen comments across multiple files.
Addressed bugs in doxygen and made spelling updates for clarity.

gloopydoop committed:
cleaning

gloopydoop committed:
cleaning

timfelle committed:
Fix segmentation fault and improve logging (#134)

Address segmentation faults by ensuring proper memory management and
initialization. Enhance logging capabilities for better debugging and
add tests to verify functionality. Align code with new checkpoint
structures and clean up unnecessary outputs.

---------

Co-authored-by: Harrison Nobis <58177854+gloopydoop@users.noreply.github.com>

gloopydoop committed:
Merge branch 'feature/clean_passive_scalar' of github.com:ExtremeFLOW/neko-top into feature/clean_passive_scalar

gloopydoop commented on sources/problems/objectives/scalar_mixing_objective_function.f90:
```diff
@@ -0,0 +1,290 @@
+    ! eg,
+    ! call field_rzero(this%sensitivity)
+    !
+    ! THIS SHOULD BE REPLACED WITH A ZERO CONTRIBUTION AFTER THE DESIGN UPDATE
```
This is referring to the fact that we should have the `simulation_t` computing the $u \cdot u^\dagger$, and this subroutine is for computing additional sensitivity, ie, like we have in the lube_term

gloopydoop committed:
merge develop and add adjoint and scalar to simulation

gloopydoop committed:
design with adjoint_fluid

gloopydoop committed:
cleaning

gloopydoop committed:
cleaning

gloopydoop committed:
cleaning

gloopydoop committed:
merge develop

gloopydoop committed:
formatting

gloopydoop committed:
dox

gloopydoop committed:
readme

gloopydoop committed:
dox

gloopydoop committed:
dox

gloopydoop committed:
dox

gloopydoop committed:
dox

gloopydoop committed:
dox

gloopydoop committed:
dox

gloopydoop committed:
dox

timfelle commented on sources/adjoint/adjoint_scalar_pnpn.f90:
```diff
@@ -0,0 +1,550 @@
+     procedure, pass(this) :: setup_bcs_ => adjoint_scalar_pnpn_setup_bcs_
+  end type adjoint_scalar_pnpn_t
+
+  interface
+     !> Boundary condition factory. Both constructs and initializes the object.
+     !! @details Will mark a mesh zone for the bc and finalize.
+     !! @param[inout] object The object to be allocated.
+     !! @param[in] scheme The `adjoint_scalar_pnpn` scheme.
+     !! @param[inout] json JSON object for initializing the bc.
+     !! @param[in] coef SEM coefficients.
+     module subroutine adjoint_bc_factory(object, scheme, json, coef, user)
```
```suggestion
  !> Boundary condition factory. Both constructs and initializes the object.
  !! @details Will mark a mesh zone for the bc and finalize.
  !! @param[inout] object The object to be allocated.
  !! @param[in] scheme The `adjoint_scalar_pnpn` scheme.
  !! @param[inout] json JSON object for initializing the bc.
  !! @param[in] coef SEM coefficients.
  interface adjoint_bc_factory
     module subroutine adjoint_bc_factory(object, scheme, json, coef, user)
```

timfelle commented on examples/passive_scalar/passive_scalar.case:
```diff
@@ -0,0 +1,189 @@
+                "weight": 1.0
+            },
+        ],
+        "constraints": [
+        ],
```
Avoid empty arrays.
```suggestion
```

timfelle commented on examples/passive_scalar/readme.md:
```diff
@@ -0,0 +1,127 @@
+# Passive scalar example (moving towards the mixer)
```
Lets give it a reference.
Please add it to the example list in [examples.md](examples/examples.md)
```suggestion
# Passive scalar example (moving towards the mixer) {#passive-scalar}
```

timfelle commented on examples/passive_scalar/readme.md:
```diff
@@ -0,0 +1,127 @@
+
+Often times one can replace an objective on a BC with a volume integral, 
+ie, instead of of evaluated the "mixedness" on the outlet surface, measure
+the mixedness in a volume downsteam close to the outlet.
```
```suggestion
the mixedness in a volume downstream close to the outlet.
```

timfelle commented on examples/passive_scalar/readme.md:
```diff
@@ -0,0 +1,127 @@
+ie, instead of of evaluated the "mixedness" on the outlet surface, measure
+the mixedness in a volume downsteam close to the outlet.
+This turns the surface integral (and subsequent adjoint BCs) into a 
+volume integral (and subseqent adjoint source terms) which may be easier to
```
```suggestion
volume integral (and subsequent adjoint source terms) which may be easier to
```

timfelle commented on examples/passive_scalar/readme.md:
```diff
@@ -0,0 +1,127 @@
+implemented a `adjoint_enhanced_mixing_scalar_source_term` instead.
+
+> I also feel the design needs to enter the passive scalar
+> equation. Probably not in the difusive term, but I think it should show up
```
```suggestion
> equation. Probably not in the diffusive term, but I think it should show up
```

timfelle commented on examples/passive_scalar/readme.md:
```diff
@@ -0,0 +1,127 @@
+- obviously testing... but we don't really have a good case to test yet
+When I did my derivation (I'll write it up neatly at some point...) 
+- Double check with Casper what the BC's are, it looks like he keeps drawing
+parabolic profiles, is there a periodic or symetric directions?
```
```suggestion
parabolic profiles, is there a periodic or symmetric directions?
```

timfelle commented on examples/passive_scalar/user.f90:
```diff
@@ -0,0 +1,103 @@
+    real(kind=rp), intent(in) :: t
+    integer, intent(in) :: tstep
+
+    ! Inflow velocity profile is a parabaloid
```
```suggestion
    ! Inflow velocity profile is a paraboloid
```

timfelle commented on sources/adjoint/adjoint_case.f90:
```diff
@@ -52,8 +59,13 @@ module adjoint_case
   !! Todo: This should Ideally be a subclass of case_t, however, that is not yet
   !! suppoerted by Neko.
```
```suggestion
  !! supported by Neko.
```

timfelle commented on sources/adjoint/adjoint_output.f90:
```diff
@@ -32,13 +32,14 @@
+  use field_list, only : field_list_t
+  use neko_config, only : NEKO_BCKND_DEVICE
+  use device, only : device_memcpy, DEVICE_TO_HOST
+  use output, only : output_t
```
```suggestion
  use num_types, only: rp
  use adjoint_fluid_scheme, only: adjoint_fluid_scheme_t
  use scalar_scheme, only: scalar_scheme_t
  use adjoint_scalar_scheme, only: adjoint_scalar_scheme_t
  use field_list, only: field_list_t
  use neko_config, only: NEKO_BCKND_DEVICE
  use device, only: device_memcpy, DEVICE_TO_HOST
  use output, only: output_t
```

timfelle commented on sources/adjoint/adjoint_scalar_convection_source_term.f90:
```diff
@@ -0,0 +1,190 @@
+  use field_math, only: field_subcol3
+  use operators, only: opgrad, grad
+  use math, only: invcol2
+  use gather_scatter, only : GS_OP_ADD
```
```suggestion
  use num_types, only: rp
  use field_list, only: field_list_t
  use field, only: field_t
  use scratch_registry, only: neko_scratch_registry
  use json_module, only: json_file
  use source_term, only: source_term_t
  use coefs, only: coef_t
  use field_math, only: field_subcol3
  use operators, only: grad
```

timfelle commented on sources/adjoint/adjoint_scalar_convection_source_term.f90:
```diff
@@ -0,0 +1,190 @@
+
+    ! So should the Brinkman term actually....
+
+    call grad(dsdx%x,dsdy%x,dsdz%x,this%s%x, this%coef)
```
Device support?

gloopydoop replied:
I'm pretty sure it's handled internally, there's a `device_get_ptr` inside of `opr_device_opgrad`

timfelle commented on sources/adjoint/adjoint_scalar_convection_source_term.f90:
Why is the other source term under source terms, where this source term is not
under source terms?

gloopydoop replied:
Yeh this is a good point. In my mind this source term is a bit strange because it is only used to couple the adjoint fluid to the adjoint scalar, so I thought it didn't really fit into standard source terms. To me it fits in more with the advective terms (it isn't a advective term... but it comes from perturbing the advective term) So ideally this would be evaluated on the dealiased mesh.  But I guess it doesn't fit with the advection terms but it also doesn't really fit here either :sweat_smile: 

I'll pop it in with the others source terms, I guess that's the most logical place.

timfelle commented on sources/source_terms/adjoint_mixing_scalar_source_term.f90:
```diff
@@ -0,0 +1,191 @@
+!> Implements the `adjoint_mixing_scalar_source_term` type.
+! this is a such a dumb name
+module adjoint_mixing_scalar_source_term
+  use num_types, only : rp
+  use field_list, only : field_list_t
+  use field, only: field_t
+  use field_registry, only: neko_field_registry
+  use scratch_registry, only: neko_scratch_registry
+  use json_module, only : json_file
+  use json_utils, only: json_get, json_get_or_default
+  use source_term, only : source_term_t
+  use coefs, only : coef_t
+  use neko_config, only : NEKO_BCKND_DEVICE
+  use utils, only : neko_error
+  use field_math, only: field_addcol3, field_add2, field_cadd, field_col2, &
+       field_add2s2, field_copy, field_cfill
+  use operators, only: opgrad
+  use mask_ops, only: mask_exterior_const, compute_masked_volume
+  use point_zone, only: point_zone_t
+  use math, only: col2, invcol2
```
Please remove the imports that are not used.
col2, invcol2, opgrad are not used and I stopped checking for more.

gloopydoop replied:
on it :saluting_face:

timfelle commented on sources/simulation/simulation.f90:
```diff
@@ -56,11 +60,14 @@ module simulation_m
      !> The fluid
-     class(fluid_scheme_incompressible_t), public, pointer :: &
-          fluid_scheme => null()
-
+     class(fluid_scheme_incompressible_t), public, pointer :: fluid => null()
+     !> The scalar
+     type(scalar_pnpn_t), public, pointer :: scalar => null()
+     !> The adjoint fluid
+     class(adjoint_fluid_scheme_t), public, pointer :: adjoint_fluid => null()
+     !> The adjoint scalar
+     type(adjoint_scalar_pnpn_t), public, pointer :: adjoint_scalar => null()
```
As you mentioned yourself, these are already part of the cases. Please don't add
more stuff which will be deleted in a week or two when we get to fixing the simulation.

gloopydoop replied:
I've actually changed my mind with this badboy!!!
It really makes other sections of the code (look at some of the objective functions for instance) much cleaner. i.e. to say `simulation%fluid` or `simulation%adjoint_fluid` etc, as opposed to `simulation%adjoint_case%fluid_adj`. It make otherwise long statements concise and legible. They're just pointers so they cost nothing, and it makes the code more user friendly from a mid-user-mid-developer perspective (ie, keen to implement additional constraints or objectives)
So I'm happy the way this is 👍

timfelle commented on sources/problems/objectives/scalar_mixing_objective_function.f90:
```diff
@@ -0,0 +1,281 @@
+!> An objective function corresponding to the mixing of a passive scalar
+!! \f$ F = \frac{1}{|\Omega_{obj}|}\int_{\Omega_{obj}}
+!! \frac{1}{2}\left(\phi - \phi_{ref}\right)^2 d\Omega, \f$
+!
```
```suggestion
```

timfelle commented on sources/adjoint/advection_adjoint.f90:
```diff
@@ -44,8 +44,7 @@ module advection_adjoint
      procedure(compute_adv_lin), pass(this), deferred :: compute_adjoint
-     ! TODO
-     ! procedure(compute_scalar_adv_lin), pass(this), deferred :: compute_scalar
+     procedure(compute_scalar_adv_lin), pass(this), deferred :: compute_adjoint_scalar
```
```suggestion
     procedure(compute_scalar_adv_lin), pass(this), deferred :: &
          compute_adjoint_scalar
```

timfelle commented on sources/adjoint/adv_adjoint_dealias.f90:
```diff
@@ -667,7 +677,90 @@ subroutine compute_linear_advection_dealias(this, vx, vy, vz, vxb, vyb, vzb, &
+
+
+
+
```
```suggestion

```

timfelle commented on sources/adjoint/adv_adjoint_dealias.f90:
```diff
@@ -667,7 +677,90 @@ subroutine compute_linear_advection_dealias(this, vx, vy, vz, vxb, vyb, vzb, &
+            call this%GLL_to_GL%map(temp, f_GL, 1, this%Xh_GLL)
+            call sub2(fs%x(idx, 1, 1, 1), temp, this%Xh_GLL%lxyz)
+
+         enddo
```
```suggestion
         end do
```

timfelle reviewed (changes requested):
*** ERROR: Invalid user defined inflow condition ***

timfelle committed:
Apply suggestions from code review

timfelle committed:
fix format

timfelle committed:
Merge branch 'develop' into dev/cleanup

timfelle committed:
fix doc

timfelle committed:
Merge commit 'f6fe84296215950df3d343027bdb9168e8c5b4a6' into feature/clean_passive_scalar

timfelle committed:
minor update

timfelle commented on sources/simulation/simulation.f90:
```diff
@@ -37,6 +37,10 @@ module simulation_m
+  use adjoint_fluid_pnpn, only: adjoint_fluid_pnpn_t
+  use scalar_pnpn, only: scalar_pnpn_t
+  use adjoint_scalar_pnpn, only: adjoint_scalar_pnpn_t
   use fluid_pnpn, only: fluid_pnpn_t
```
Please go through and remove old imports. Something like the fluid_pnpn_t is not
used

gloopydoop replied:
I think it's needed when you go 

```fortran
select type (fluid => this%neko_case%fluid)                                 
    type is (fluid_pnpn_t)                                                      
       this%fluid => fluid                                                      
                                                                                
    end select
```

timfelle committed:
Merge remote-tracking branch 'origin/develop' into feature/clean_passive_scalar

gloopydoop committed:
cleaning mixing source term

gloopydoop committed:
moved convection source term

gloopydoop committed:
woops

timfelle reviewed (changes requested):
I still get *** ERROR: Invalid user defined inflow condition ***, could you look into it?
```
./setup.sh -c
./run.sh -rc passive_scalar
```

gloopydoop committed:
scalar case

gloopydoop committed:
example case file

gloopydoop commented:
> I still get *** ERROR: Invalid user defined inflow condition ***, could you look into it?
> 
> ```
> ./setup.sh -c
> ./run.sh -rc passive_scalar
> ```

I just merged in the `.case` file from #136 (which I think makes that PR obsolete now, since we have more or less everything else in the case file now :ok_hand: )

It runs on my end, if you still see that error let's schedule a call or something because that's definitely something to look into.

timfelle reviewed (approved):
Let's get it in and fix the example after, since #1803 on the neko side will break stuff again. We can work on pr #136 for the fix of the example
